### ISSUE TITLE
issue: useFieldArray - old values appear after remove() and append()

### ISSUE THREAD

**Issue by zirkelc** [react-hook-form/react-hook-form#7137]
### Version Number

7.20.2

### Codesandbox/Expo snack

https://codesandbox.io/s/usefieldarray-lbul1?file=/src/App.tsx

### Steps to reproduce

I'm using RHF useFieldAray and display the fields in Material-UI DataGrid. The initial values are being set via reset(). Then I remove all rows from the DataGrid with remove() and then append() a new empty row. Then, I double-click into the first cell to toggle edit mode and the old values which were deleted appear again.

1. Open Sandbox
2. Select all rows in table
3. Click 'Remove'
4. Click 'Add'
5. Double-click into first row and column to enable edit mode
6. Old values **Key1** and **Name1** appear again

### Expected behaviour

append() should add an empty row without values.

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
we don't support the following usage: 

```
const handleRemoveProperty = () => {
  const indexes = selectedRows.map((rowId) =>
    rows.findIndex((row) => row.id === rowId)
  );
  remove(indexes);
};
```

please refer to the doc for proper usage: 

https://react-hook-form.com/api/usefieldarray

**Comment by zirkelc** [comment]
Could you please explain briefly what is wrong with this usage?

**Comment by vcarel** [comment]
Hi @bluebill1049, I'm calling `remove(index)` to remove a row when clicking on a delete button.

![image](https://user-images.githubusercontent.com/1541093/149953626-48bd8e09-1709-4be7-9926-28689ebac974.png)

Every-time I remove a row, there's obviously a new render. However, on the 2nd remove, the last row of the table is duplicated (or it's revived when it was the deleted row).

**Comment by bluebill1049** [comment]
hey, @vcarel it doesn't sound right to me. Can you share some of your code? I am happy to look into it. I am worried the usage is incorrect, although it's "working" with replace, it may cause other issues. feel free to send me a message at twitter as well: https://twitter.com/bluebill1049

**Comment by vcarel** [comment]
Hi @bluebill1049, I was writing a long explanation but you've been faster.

TLDR; this was my mistake (how surprising 😅)... but let me explain! Maybe it will help other people running into the same issue.

…

So, I have a table containing input fields, that I can sort by drag'n drop:

![image](https://user-images.githubusercontent.com/1541093/150024578-c25f9082-0792-4982-b73e-f861cf3fce0d.png)

I'm using react-dnd for that purpose, which works great! Thanks to this lib, when I drag a row, I'm able to call a callback to the parent table with the "from" and "to" indexes. Just as the arguments of the `useFieldsArray` `move` function BTW.

Unfortunately, for a silly reason, calling the `move` function while dragging a row wouldn't work.

**That's the root cause of my issue ⬇️**

So instead of calling `move`, I used an intermediate state to reorder the rows on the fly. My data structure was as follows:

```
const { fields } = useFieldArray(…)
const rows = useMemo(() => fields.map((field, fieldIndex) => ({ field, fieldIndex })), [fields])
const [orderedRows, setOrderedRows] = useState(rows)
```

My table was then based on `orderedRows`.

In this solution, when I drag a row, I just call `setOrderedRows` with the new order. Leaving the `fields` unchanged (I store the `fieldIndex` in my data structure in that purpose)

Besides, to update the table with any changes on the `fields`, I added the following useEffect:

```
  useEffect(() => {
    setOrderedRows(rows)
  }, [rows])
```

... which is what prevented `remove` to work properly in the end.

Why? Let's see what happens when I call the `useFieldsArray`'s `remove` function:

1. First, react-hook-form removes the specified field, as expected.
2. Then the component is re-rendered, because the state has changed internally. Again, this is 100% expected.
3. Then my table is re-rendered based on the `orderRows`, which is still the "old" version at this time (the useEffect is executed _after_ the render) 😬

... meaning the old row is re-rendered one more time... causing it to be registered like a fresh new row in react-hook-form 😱

👻 Mystery solved! 👻

**Comment by bluebill1049** [comment]
That's awesome! @vcarel I don't even know it works with react-dnd... awesome you sort it out 👯‍♀️ Happy Wednesday :)
